### ISSUE TITLE
Keycloak 19.0.1 Kubernetes liveness/readiness probes intermittently return HTTP/500

### ISSUE THREAD

**Issue by sbonds** [keycloak/keycloak#15799]
### Area

core

### Describe the bug

When running Keycloak 19.0.1 in a Kubernetes cluster the health checks intermittently (1-2% of the time on a 10s probe) return an `HTTP/1.1 500 Internal Server Error` with zero content instead of the expected `HTTP/200` with JSON.

### Version

19.0.1

### Expected behavior

The probe should not fail while Keycloak is healthy (e.g. the HTTP 500 errors are not expected.)

When a probe fails it should return JSON describing what component failed.

### Actual behavior

Approximately 1-2% of the time under normal operation I will see the Kubernetes probes fail.

### How to Reproduce?

Set a single keycloak 19.0.1 container running with these liveness settings:

```
delay=300s
timeout=5s
period=10s
success=1
failure=3
```

Enable debugging via these environment variables:

```
QUARKUS_LOG_MIN_LEVEL = "DEBUG"
QUARKUS_LOG_LEVEL = "DEBUG"
QUARKUS_HTTP_ACCESS_LOG_ENABLED = "true"
QUARKUS_HTTP_ACCESS_LOG_PATTERN = "%{DATE_TIME}\n%{REQUEST_LINE}\n%{ALL_REQUEST_HEADERS}\nresponse code: %{RESPONSE_CODE}"
```

Check the Kubernetes events for failed probes and confirm the 500 response via the Quarkus logs.

## Successful debug log

Here is an example of a successful probe with this level of debug logging. The timestamps were trimmed to make it narrower and for easier comparison:

```
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) new JtaTransactionWrapper
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) was existing? false
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) JtaTransactionWrapper  commit
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) JtaTransactionWrapper end
DEBUG [org.keycloak.services.scheduled.ScheduledTaskRunner] (Timer-0) Executed scheduled task AbstractLastSessionRefreshStoreFactory$$Lambda$1219/0x0000000840b2e440
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) new JtaTransactionWrapper
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) was existing? false
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) JtaTransactionWrapper  commit
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) JtaTransactionWrapper end
DEBUG [org.keycloak.services.scheduled.ScheduledTaskRunner] (Timer-0) Executed scheduled task AbstractLastSessionRefreshStoreFactory$$Lambda$1219/0x0000000840b2e440
DEBUG [io.quarkus.vertx.http.runtime.ForwardedParser] (vert.x-eventloop-thread-0) Recalculated absoluteURI to http://10.42.214.197:8080/auth/health/live
DEBUG [io.quarkus.vertx.http.runtime.ForwardedParser] (vert.x-eventloop-thread-1) Recalculated absoluteURI to http://10.42.214.197:8080/auth/health/ready
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-2) new JtaTransactionWrapper
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-3) new JtaTransactionWrapper
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-2) was existing? false
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-3) was existing? false
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-3) JtaTransactionWrapper  commit
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-2) JtaTransactionWrapper  commit
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-3) JtaTransactionWrapper end
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-2) JtaTransactionWrapper end
INFO  [io.quarkus.http.access-log] (executor-thread-1)
GET /auth/health/ready HTTP/1.1
User-Agent: kube-probe/1.22
Accept: */*
Connection: close
host: 10.42.214.197:8080
200
[io.quarkus.http.access-log] (executor-thread-0)
GET /auth/health/live HTTP/1.1
User-Agent: kube-probe/1.22
Accept: */*
Connection: close
host: 10.42.214.197:8080
200
```

## Failure debug log

Here's an example of the same level of logging from when a 500 error was returned:

```
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) new JtaTransactionWrapper
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) was existing? false
DEBUG [org.keycloak.models.sessions.infinispan.changes.sessions.PersisterLastSessionRefreshStore] (Timer-0) Updating 0 userSessions with lastSessionRefresh: 1669762938
DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (Timer-0) `hibernate.connection.provider_disables_autocommit` was enabled.  This setting should only be enabled when you are certain that the Connections given to Hibernate by the ConnectionProvider have auto-commit disabled.  Enabling this setting when the Connections do not have auto-commit disabled will lead to Hibernate executing SQL operations outside of any JDBC/SQL transaction.
DEBUG [org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl] (Timer-0) Hibernate RegisteredSynchronization successfully registered with JTA platform
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) JtaTransactionWrapper  commit
DEBUG [org.hibernate.engine.transaction.internal.TransactionImpl] (Timer-0) On TransactionImpl creation, JpaCompliance#isJpaTransactionComplianceEnabled == false
DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (Timer-0) Initiating JDBC connection release from beforeTransactionCompletion
DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (Timer-0) Initiating JDBC connection release from afterTransaction
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) JtaTransactionWrapper end
DEBUG [org.keycloak.services.scheduled.ScheduledTaskRunner] (Timer-0) Executed scheduled task AbstractLastSessionRefreshStoreFactory$$Lambda$1219/0x0000000840b2e440
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) new JtaTransactionWrapper
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) was existing? false
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) JtaTransactionWrapper  commit
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (Timer-0) JtaTransactionWrapper end
DEBUG [org.keycloak.services.scheduled.ScheduledTaskRunner] (Timer-0) Executed scheduled task AbstractLastSessionRefreshStoreFactory$$Lambda$1219/0x0000000840b2e440
DEBUG [io.quarkus.vertx.http.runtime.ForwardedParser] (vert.x-eventloop-thread-0) Recalculated absoluteURI to http://10.42.214.197:8080/auth/health/live
DEBUG [io.quarkus.vertx.http.runtime.ForwardedParser] (vert.x-eventloop-thread-1) Recalculated absoluteURI to http://10.42.214.197:8080/auth/health/ready
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) new JtaTransactionWrapper
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-0) new JtaTransactionWrapper
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) was existing? false
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-0) was existing? false
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-3) JtaTransactionWrapper  commit
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-2) JtaTransactionWrapper  commit
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-3) JtaTransactionWrapper end
DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-2) JtaTransactionWrapper end
INFO  [io.quarkus.http.access-log] (executor-thread-2)
GET /auth/health/live HTTP/1.1
User-Agent: kube-probe/1.22
Accept: */*
Connection: close
host: 10.42.214.197:8080
500
INFO  [io.quarkus.http.access-log] (executor-thread-3)
GET /auth/health/ready HTTP/1.1
User-Agent: kube-probe/1.22
Accept: */*
Connection: close
host: 10.42.214.197:8080
500
```

## Differences

The main difference are these lines:

```
DEBUG [org.keycloak.models.sessions.infinispan.changes.sessions.PersisterLastSessionRefreshStore] (Timer-0) Updating 0 userSessions with lastSessionRefresh: 1669762938
DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (Timer-0) `hibernate.connection.provider_disables_autocommit` was enabled.  This setting should only be enabled when you are certain that the Connections given to Hibernate by the ConnectionProvider have auto-commit disabled.  Enabling this setting when the Connections do not have auto-commit disabled will lead to Hibernate executing SQL operations outside of any JDBC/SQL transaction.
DEBUG [org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl] (Timer-0) Hibernate RegisteredSynchronization successfully registered with JTA platform
```

and

```
DEBUG [org.hibernate.engine.transaction.internal.TransactionImpl] (Timer-0) On TransactionImpl creation, JpaCompliance#isJpaTransactionComplianceEnabled == false
DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (Timer-0) Initiating JDBC connection release from beforeTransactionCompletion
DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (Timer-0) Initiating JDBC connection release from afterTransaction
```

Perhaps there's a race condition where if the status gets probed right when `PersisterLastSessionRefreshStore` is active the probe fails?

**Comment by souravs17031999** [comment]
Hi @sbonds , I was facing similar issue in K8S and it was discussed earlier here, the fix is released in ```19.0.2```

https://github.com/keycloak/keycloak/issues/13937 (last few comments)
https://github.com/keycloak/keycloak/issues/14011

**Comment by mposolda** [comment]
@souravs17031999 Thanks for your comment!

Closing as this is probably fixed in latest Keycloak main. Please re-open just if still seeing this in the newest Keycloak 21 version
